### PR TITLE
Sparkc 308

### PR DESCRIPTION
**Pull request by RussellSpitzer**

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+ * Adds the ability to add additional Predicate Pushdown Ruels at Runtime (SPARKC-308)
```
Comment (jacek-lewandowski) [inline CHANGES.txt]: typo: Ruels -> Rules

File: doc/reference.md [hunk 2]
```diff
@@ -93,6 +93,13 @@ may also be used. ("127.0.0.1,192.168.0.1")
 <table class="table">
 <tr><th>Property Name</th><th>Default</th><th>Description</th></tr>
 <tr>
+  <td><code>sql.pushdown.additionalclasses</code></td>
```
Comment (jacek-lewandowski) [inline doc/reference.md]: additionalClasses ?

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala [hunk 3]
```diff
@@ -91,19 +91,52 @@ private[cassandra] class CassandraSourceRelation(
   def buildScan(): RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
 
   override def unhandledFilters(filters: Array[Filter]): Array[Filter] = filterPushdown match {
-    case true =>
-      (filters.toSet -- new PredicatePushDown(filters.toSet, tableDef).predicatesToPushDown).toArray
-    case fasle => filters
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala]: wow, good catch!

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraPredicateRules.scala [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  handledByCassandra: Set[Filter],
+  handledBySpark: Set[Filter] ){
+  override def toString(): String = {
+    s"""C* Filters: [${handledByCassandra.mkString(", ")}]
+       |Spark Filters [${handledBySpark.mkString(", ")}]""".stripMargin
+  }
+}
+
+trait CassandraPredicateRules{
+  def applyRules(predicates: AnalyzedPredicates, tableDef: TableDef): AnalyzedPredicates
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraPredicateRules.scala]: maybe just `apply`? woudl be `someCassandraPredicateRule.apply(...)`, or just `someCassandraPredicateRules(...)`

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala [hunk 5]
```diff
@@ -91,19 +91,52 @@ private[cassandra] class CassandraSourceRelation(
...
+    val sc = sqlContext.sparkContext
+
+    /* So we can set this in testing to different values without
+     making a new context check local property as well */
+    val userClasses: Option[String] =
+      sc.getConf.getOption(AdditionalCassandraPushDownRulesParam.name)
+        .orElse(Option(sc.getLocalProperty(AdditionalCassandraPushDownRulesParam.name)))
+
+    userClasses match {
+      case Some(classes) => classes.split(",").map(ReflectionUtil.findGlobalObject[CassandraPredicateRules]).reverse
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala]: would be safer to do `classes.trim.split("""\s*,\s*""")`

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala [hunk 6]
```diff
@@ -91,19 +91,52 @@ private[cassandra] class CassandraSourceRelation(
...
+      case None => AdditionalCassandraPushDownRulesParam.default
+    }
+  }
+
+  private def predicatePushDown(filters: Array[Filter]) = {
+    logInfo(s"Input Predicates: [${filters.mkString(", ")}]")
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala]: please don't `logInfo` so much :)
I'd make those messages at `DEBUG` level.

**Comment by jacek-lewandowski** [comment]
Few minor comments. Other than that, LGTM +1

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala (hunk 6 above)
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala]: I like logging :dancer: but ok

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala (hunk 3 above)
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala]: I feel a slight sadness that `case fasle` didn't throw a compiler error :(

**Comment by jacek-lewandowski** [comment]
btw. I just realised - this is probably unneeded to reverse user classes list and then doing foldRight - you can just do foldLeft

**Pull request merged**